### PR TITLE
fix(resolver): De-prioritize no-rust-version in MSRV resolver

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This is a corner case without a good answer.
As such, this change leans on some happy-path entries existing and
preferring those.

### Additional information

This was originally discussed around the time of #12950 but was held off.

When working on this, I was considering other heuristics like
- If a future version has an MSRV, assume that it applies also to the current version
  - This can be added in the future
  - We likely would want to consider an alternative value, like inferring the rust-version from the manifest or the rust-version used from publish
- Sort no-MSRV versions of a package by minimal versions
  - The lower the version, the more likely it is to be compatible
  - This likely could apply to incompatible MSRVs (or we could reverse-sort those by rust-version) but those will error anyways without `--ignore-rust-version`, so I decided against these
  - I realized this was a backdoor to minimal versions for dependencies without a MSRV and that the community support isn't there for that yet to be a high enough quality of an experience

### CODE REVIEW

File: src/cargo/core/resolver/version_prefs.rs [hunk 1]
```diff
@@ -84,7 +84,38 @@ impl VersionPreferences {
...
+                        if a <= max_rust_version {
+                            return Ordering::Less;
+                        } else {
+                            return Ordering::Greater;
+                        }
+                    }
+                }
                 let msrv_a = a.rust_version() <= self.max_rust_version.as_ref();
                 let msrv_b = b.rust_version() <= self.max_rust_version.as_ref();
                 let msrv_cmp = msrv_a.cmp(&msrv_b).reverse();
```
Comment (Eh2406) [inline src/cargo/core/resolver/version_prefs.rs:121]: Do we still need these comparisons?

File: src/cargo/core/resolver/version_prefs.rs [hunk 2]
```diff
@@ -84,7 +84,38 @@ impl VersionPreferences {
                 return previous_cmp;
             }
 
-            if self.max_rust_version.is_some() {
+            if let Some(max_rust_version) = &self.max_rust_version {
+                match (a.rust_version(), b.rust_version()) {
+                    // Fallback
+                    (None, None) => {}
+                    (Some(a), Some(b)) if a == b => {}
```
Comment (Eh2406) [inline src/cargo/core/resolver/version_prefs.rs:91]: Can be combined with the previous as `(a, b) if a == b =>`

Comment (epage) [reply src/cargo/core/resolver/version_prefs.rs:91]: The compiler doesn't recognize that `None == None` and complains about the `match` being non-exhaustive

**Pull request merged**